### ISSUE TITLE
Bug: [ban-types] Applying suggested fix results into another error

### ISSUE THREAD

**Issue by phalgunv** [typescript-eslint/typescript-eslint#8090]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.3.2&fileType=.ts&code=IYIwzgLgTsDGEAJYBthjAgQmgpgg3gLABQCCokM8CA9gHYDCyNYOAPACoB8AFACbAIwAFwIOASlEA3GgEs%2BAbhIBfEiRRoMDABazkfBDgAeEHHT4ZsrAiTL0mLdpho1kOYHV4Cho56-d04jbEtgiqxMpAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6AIwEMnajTUMkRNGgB7aJHBgAviGlA&tsconfig=&tokens=false

### Repro Code

```TypeScript
abstract class Base {
  abstract onClose<T>(data: T): void;
}

class Child extends Base {
  onClose<Boolean>(data: Boolean) {

  }
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
     "@typescript-eslint/ban-types": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

Applying fix for the error should not result in another error.

### Actual Result

Before fix:
![image](https://github.com/typescript-eslint/typescript-eslint/assets/10145093/f71bba50-7b5f-45a2-83f5-15cbe77f719e)

After fix:
There is a new error (TS2416) as a result of applying the suggested fix.
![image](https://github.com/typescript-eslint/typescript-eslint/assets/10145093/7da2402b-146a-481e-bf0d-e8b7f86a6e53)

**Comment by phalgunv** [comment]
Updated alt text for image.

**Comment by auvred** [comment]
Probably duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/5505, https://github.com/typescript-eslint/typescript-eslint/issues/6262

**Comment by bradzacher** [comment]
Yup - in particular #6262
